pydantic/main.py

```diff
@@ -638,33 +626,39 @@ def __iter__(self) -> 'TupleGenerator':
         """
         so `dict(model)` works
         """
-        yield from self._iter()
+        yield from self.__dict__.items()
 
     def _iter(
         self,
         to_dict: bool = False,
         by_alias: bool = False,
-        allowed_keys: Optional['SetStr'] = None,
         include: Union['AbstractSetIntStr', 'DictIntStrAny'] = None,
         exclude: Union['AbstractSetIntStr', 'DictIntStrAny'] = None,
         exclude_unset: bool = False,
         exclude_defaults: bool = False,
         exclude_none: bool = False,
     ) -> 'TupleGenerator':
 
+        allowed_keys = self._calculate_keys(include=include, exclude=exclude, exclude_unset=exclude_unset)
+        if allowed_keys is None and not (to_dict or by_alias or exclude_unset or exclude_defaults or exclude_none):
+            # huge boost for plain _iter()
+            yield from self.__dict__.items()
+            return
+
         value_exclude = ValueItems(self, exclude) if exclude else None
         value_include = ValueItems(self, include) if include else None
 
-        if exclude_defaults:
-            if allowed_keys is None:
-                allowed_keys = set(self.__fields__)
-            for k, v in self.__field_defaults__.items():
-                if self.__dict__[k] == v:
-                    allowed_keys.discard(k)
-
-        for k, v in self.__dict__.items():
-            if allowed_keys is None or k in allowed_keys:
-                value = self._get_value(
+        for k, v in self.__iter__():
+            if (
+                (allowed_keys is not None and k not in allowed_keys)
+                or (exclude_none and v is None)
+                or (exclude_defaults and self.__field_defaults__.get(k, _missing) == v)
+            ):
+                continue
+            if by_alias:
```

Comment:
```suggestion
            if by_alias and k in self.__fields__:
                k = self.__fields__[k].alias
```

Reply:
or perhaps
```suggestion
            if by_alias:
                try:
                    k = self.__fields__[k].alias
                except KeyError:
                    pass
```
Would be faster as you'd only need to do one look up of `self.__fields`

Reply:
Catching exceptions is actually very slow, so it will be much slower if k is not in fields and slightly slower otherwise:
```py
In [1]: d = {str(i): i for i in range(1000)}
In [2]: %%timeit
   ...: try:
   ...:     k = d[1]
   ...: except KeyError:
   ...:     pass
   ...:
218 ns ± 1.83 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %%timeit
   ...: if 1 in d:
   ...:     k = d[1]
   ...:
32.8 ns ± 0.201 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %%timeit
   ...: try:
   ...:     k = d["1"]
   ...: except KeyError:
   ...:     pass
   ...:
52.7 ns ± 0.43 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [5]: %%timeit
   ...: if 1 in d:
   ...:     k = d["1"]
   ...:
32.7 ns ± 0.303 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

So I think, better stick to first approach.

Reply:
ok